Support mesos framework authentication.

everpeace opened the pull request:
This is for #22 .
I'm pleased if I could have some feedbacks.

adam-mesos commented on src/main/java/com/ebay/myriad/scheduler/MyriadDriver.java:
```diff
@@ -17,17 +17,23 @@
 import org.apache.mesos.Protos.FrameworkID;
 import org.apache.mesos.Protos.FrameworkInfo;
 import org.apache.mesos.Protos.FrameworkInfo.Builder;
+import org.apache.mesos.Protos.Credential;
```
Please alphabetize, up before FrameworkID

adam-mesos commented on src/main/java/com/ebay/myriad/scheduler/MyriadDriver.java:
```diff
@@ -56,8 +62,29 @@ public MyriadDriver(final MyriadScheduler scheduler,
+        if (StringUtils.isNotEmpty(mesosAuthenticationPrincipal)) {
+            frameworkInfoBuilder.setPrincipal(mesosAuthenticationPrincipal);
+
+            Credential.Builder credBuilder = Credential.newBuilder();
```
Would rather see this called "credentialBuilder" instead of "credBuilder", but I wouldn't hold back the merge for it.

adam-mesos commented on src/main/java/com/ebay/myriad/scheduler/MyriadDriver.java:
```diff
@@ -56,8 +62,29 @@ public MyriadDriver(final MyriadScheduler scheduler,
+                try {
+                    credBuilder.setSecret(ByteString.readFrom(new FileInputStream(mesosAuthenticationSecretFilename)));
+                } catch (FileNotFoundException ex) {
+                    LOGGER.error("mesos authentication secret file was not found", ex);
```
Please capitalize error messages (and ideally capitalize "Mesos" in general)

adam-mesos commented on src/main/java/com/ebay/myriad/scheduler/MyriadDriver.java:
```diff
@@ -56,8 +62,29 @@ public MyriadDriver(final MyriadScheduler scheduler,
+                    LOGGER.error("mesos authentication secret file was not found", ex);
+                    throw new RuntimeException(ex);
+                } catch (IOException ex) {
+                    LOGGER.error("error in reading mesos authentication secret file", ex);
```
s/error/Error/
s/mesos/Mesos/
s/in // (remove 'in ')

adam-mesos commented:
Looks great @everpeace :+1: ! Thanks for the patch. If you fix the minor issues I mentioned, I'll merge this in. If you don't have time, I could fix them myself.

everpeace commented:
@adam-mesos Thanks for your feedback :-)  I'll fix them up tonight.  Are these commits better to be squashed??  I'm not sure that squashing would break this pull request entry or not.

adam-mesos commented:
Thanks! A squash would be great.

everpeace commented:
@adam-mesos I modified `MyriadDriver.java` according your feedback.  And I squashed java source code related commits and rebased to the latest `phase1` branch.  If you wanted me to squash all these commits into one commit, please tell me.

adam-mesos commented:
Looks great! How did you test this? @everpeace

everpeace commented:
@adam-mesos  Actually, I just manually confirmed `MyriadDriver` constructs `Credential` message and set `principal` in `FrameworkInfo` correctly, but I have been stacked on actual authentication part.  I am really glad if I could have some guidance for testing actual authentication.

credentials file for mesos-master:

``` sh
root@localhost$ touch /etc/mesos-master/authenticate
root@localhost$ echo -n 'mesos rocks' > /root/passwd
root@localhost$ chmod 600 /root/passwd
root@localhost$ echo -n '/root/passwd' > /etc/mesos-master/credentials
root@localhost$ restart mesos
# Then, It looks mesos-master reads the credentials file successfully.
root@localhost$ grep '/root/passwd' /var/log/syslog
Mar  4 09:17:18 vagrant-ubuntu-trusty-64 mesos-master[14335]: I0304 09:17:18.070652 14335 credentials.hpp:36] Loading credentials for authentication from '/root/passwd'
```

secret file for myriad framework:

```
hduser@localhost$ echo -n 'rocks' > /home/hduser/myriad-secret
hduser@localhost$ chmod 600 /home/hduser/myriad-secret
```

After that, due to #54, I rebuild myriad scheduler with modified `myriad-config-default.yml` below and copy myriad framework jar to `$YARN_HOME/shar/hadoop/yarn/lib` directory.

```
mesosMaster: zk://localhost:2181/mesos
... 
mesosAuthenticationPrincipal: mesos
mesosAuthenticationSecretFilename: /home/hduser/myriad-secret
```

However,  `yarn-daemon.sh` fails framework authentication like below:

```
hduser@localhost$ yarn-daemon.sh start resourcemanager
starting resourcemanager, logging to /usr/local/hadoop/logs/yarn-hduser-resourcemanager-vagrant-ubuntu-trusty-64.out
hduser@localhost$ less $YARN_HOME/logs/yarn-hduser-resourcemanager-vagrant-ubuntu-trusty-64.out
...
2015-03-04 09:50:39,709:15666(0x7fc4b8c84700):ZOO_INFO@check_events@1703: initiated connection to server [127.0.0.1:2181]
2015-03-04 09:50:39,711:15666(0x7fc4b8c84700):ZOO_INFO@check_events@1750: session establishment complete on server [127.0.0.1:2181], sessionId=0x14be38af92b002f, negotiated timeout=10000
I0304 09:50:39.715373 15712 group.cpp:313] Group process (group(1)@10.0.2.15:38233) connected to ZooKeeper
I0304 09:50:39.715420 15712 group.cpp:790] Syncing group operations: queue size (joins, cancels, datas) = (0, 0, 0)
I0304 09:50:39.715441 15712 group.cpp:385] Trying to create path '/mesos' in ZooKeeper
I0304 09:50:39.725257 15707 detector.cpp:138] Detected a new leader: (id='7')
I0304 09:50:39.725442 15712 group.cpp:659] Trying to get '/mesos/info_0000000007' in ZooKeeper
I0304 09:50:39.730058 15707 detector.cpp:433] A new leading master (UPID=master@10.0.2.15:5050) is detected
I0304 09:50:39.730132 15707 sched.cpp:234] New master detected at master@10.0.2.15:5050
I0304 09:50:39.730305 15707 sched.cpp:284] Authenticating with master master@10.0.2.15:5050
I0304 09:50:39.730417 15707 authenticatee.hpp:109] Initializing client SASL
I0304 09:50:39.734380 15707 authenticatee.hpp:133] Creating new client SASL connection
I0304 09:50:39.736425 15706 authenticatee.hpp:224] Received SASL authentication mechanisms: CRAM-MD5
I0304 09:50:39.736471 15706 authenticatee.hpp:250] Attempting to authenticate with mechanism 'CRAM-MD5'
I0304 09:50:39.737449 15706 authenticatee.hpp:270] Received SASL authentication step
E0304 09:50:39.739164 15706 sched.cpp:353] Master master@10.0.2.15:5050 refused authentication
I0304 09:50:39.739183 15706 sched.cpp:736] Got error 'Master refused authentication'
I0304 09:50:39.739197 15706 sched.cpp:1320] Asked to abort the driver
I0304 09:50:39.740515 15706 sched.cpp:777] Aborting framework ''
...
```

everpeace commented:
@adam-mesos  Sorry for delay.  However, I finally fixed framework authentication code by referring mesosphere/marathon#1290 .  I confirmed this works manually.  I squashed and rebased this branch to the latest `phase1`.  So this should be easy to merge.

mesos-master credentials:

```
mesos rocks
```

`myriad-config-default.yml`:

```
...
mesosAuthenticationPrincipal: mesos
mesosAuthenticationSecretFilename: /home/hduser/myriad-secret
```

`/home/hduser/myriad-secret`:

```
rocks
```

extracted myriad log `yarn-hduser-resourcemanager-vagrant-ubuntu-trusty-64.out` (relevant part only):

```
...
I0311 17:24:16.657812  2106 detector.cpp:433] A new leading master (UPID=master@10.0.2.15:5050) is detected
I0311 17:24:16.657902  2106 sched.cpp:234] New master detected at master@10.0.2.15:5050
I0311 17:24:16.658102  2106 sched.cpp:284] Authenticating with master master@10.0.2.15:5050
I0311 17:24:16.658217  2106 authenticatee.hpp:109] Initializing client SASL
I0311 17:24:16.684101  2106 authenticatee.hpp:133] Creating new client SASL connection
I0311 17:24:16.687476  2110 authenticatee.hpp:224] Received SASL authentication mechanisms: CRAM-MD5
I0311 17:24:16.687510  2110 authenticatee.hpp:250] Attempting to authenticate with mechanism 'CRAM-MD5'
I0311 17:24:16.688473  2110 authenticatee.hpp:270] Received SASL authentication step
I0311 17:24:16.689245  2110 authenticatee.hpp:310] Authentication success
I0311 17:24:16.689297  2110 sched.cpp:358] Successfully authenticated with master master@10.0.2.15:5050
I0311 17:24:16.691205  2110 sched.cpp:408] Framework registered with 20150304-102322-251789322-5050-17057-0000
...
```

everpeace commented:
I'm not sure exactly what went wrong though, all 'echo -n'-ed credential and secret files worked fine on the previous version of my code which read all bytes from specified secret file.  So I rolled back my branch to the previous one.

adam-mesos commented:
@everpeace Is this PR in its current state tested and ready for review?

everpeace commented:
@adam-mesos Yes, This branch was tested and ready for review!

adam-mesos commented:
Sorry it took me so long to review again. Looks great! :+1:
